gcapi/create_strategies.py

```diff
@@ -0,0 +1,455 @@
+                "display_set": self.parent.pk,
+                "interface": self.socket.slug,
+            }
+        elif isinstance(self.parent, (ArchiveItem, ArchiveItemPost)):
+            upload_session_data = {
+                "archive_item": self.parent.pk,
+                "interface": self.socket.slug,
+            }
+        else:
+            raise NotImplementedError(
+                f"{type(self.parent)} not supported for uploading images"
+            )
+
+        uploads = []
+        for file in self.content:
+            with open(file, "rb") as f:
+                uploads.append(
+                    (
+                        yield from self.hybrid_client_api.uploads.upload_fileobj(
+                            fileobj=f, filename=file.name
+                        )
+                    )
+                )
+        raw_image_upload_session = (
+            yield from self.hybrid_client_api.raw_image_upload_sessions.create(
+                uploads=[u.api_url for u in uploads],
+                **upload_session_data,
+            )
+        )
+
+        if self.parent is None:
+            post_request.upload_session = raw_image_upload_session.api_url
+            return post_request
+        else:
+            return Empty
+
+
+class ValueSocketValueCreateStrategy(SocketValueCreateStrategy):
+
+    def prepare(self):
+        yield from super().prepare()
+
+        if isinstance(self.source, HyperlinkedComponentInterfaceValue):
+            self._prepare_from_existing_socket_value(socket_value=self.source)
+        else:
+            try:
+                self._prepare_from_local_file(file=self.source)
+            except FileNotFoundError:
+                self._prepare_from_provided_value(value=self.source)
+
+    def _prepare_from_local_file(self, file):
+        cleaned_sources = clean_file_source(file, maximum_number=1)
+        self.content = cleaned_sources[0]
+
+        # Test load
+        with open(self.content) as fp:
+            json.load(fp)
+
+    def _prepare_from_existing_socket_value(self, socket_value):
+        socket = socket_value.interface
+        if socket.super_kind.casefold() == "file":
+            pass  # Nothing to do here, but we accept it
+        elif socket.super_kind.casefold() == "value":
+            self.content = io.StringIO(json.dumps(socket_value.value))
+        else:
+            raise NotImplementedError(
+                f"Socket super_kind {socket.super_kind} not supported as source"
+            )
+
+    def _prepare_from_provided_value(self, value):
+        json.dumps(value)  # Check if it is JSON serializable
+        self.content = value
+
+    def __call__(self):
+        post_request = yield from super().__call__()
+
+        if (
+            isinstance(self.source, HyperlinkedComponentInterfaceValue)
+            and self.source.interface.super_kind.casefold() == "file"
+        ):
+            temp_file = yield from _download_file(
```

Comment:
I am not really proud of this construct. I think creating a contextmanagers might work and maybe even a buffer-to-buffer where you stream the download directly into the upload.

However, with the hybrid async/sync attributes it becomes a pain to figure out who to create a contextmanager while still being able to support both async and sync.

Core idea here is that it happens by downloading things to disk first, and then streaming from there to prevent a a memory bottleneck (at least for the pure FileSocketValueStrategy). `ValuesSocketValueStrategy` (i.e. when targeting a 'value' kind socket) will generaly not run into memory problems.

Open for suggestions but atm good enough, imho.

Reply:
Note that these parts (i.e. strategy calling) are not tested here. Rather they are part of the integration testing of the next PR. This is done intentionally as to keep this PR within reviewable size. 

I had some quick-and-dirty test scripts running while developing this.

Reply:
I don't have a better suggestion at the moment, and think this is indeed fine.